fix: restrict integer variable bounds for Choco Solver

RehanChalana opened the pull request:
- Restricted integer variable bounds to Choco-recommended safe range (Integer.MIN_VALUE / 100 to Integer.MAX_VALUE / 100) to prevent internal arithmetic overflows.

- Added detailed runtime exception if solver returns null, including time limit and problem state for debugging.

RehanChalana committed:
fix: restrict integer variable bounds for Choco Solver

- Restricted integer variable bounds to Choco-recommended safe range (Integer.MIN_VALUE / 100 to Integer.MAX_VALUE / 100) to prevent internal arithmetic overflows.

- Added detailed runtime exception if solver returns null, including time limit and problem state for debugging.

RehanChalana committed:
fix: Raise IllegalArgumentException in makeIntVar when bounds exceed Choco's safe range to prevent silent overflows

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/numeric/solvers/ProblemChoco.java:
```diff
@@ -57,7 +57,14 @@ public ProblemChoco() {
 
 	public IntDomainVar makeIntVar(String name, long min, long max) {
-		assert(min>=Integer.MIN_VALUE && max<=Integer.MAX_VALUE);
+		// Choco recommends staying within Integer.MIN_VALUE / 100 and Integer.MAX_VALUE / 100
```
Please add the reference in the comment here, where choco documentation recommends that.
Also, please add a note that choco is best used to check for unsat, i.e., safety properties, but not for sat, i.e., verification properties, due to its limited integer range.

sohah reviewed (commented):
Please add to your pull request a test case, a java program along with the jpf file for which we can test your changes. Please do that for all the changes you make. These tests should be under examples folder.

RehanChalana committed:
test: Add TestChocoIntRangeAndTimeout to observer choco's timeout and bounded int range behaviour

RehanChalana commented:
@sohah I have added the source reference, Its a PDF, so I was not able to point to a specific page number.

I have also added an example: TestChocoIntRangeAndTimeout. In this test, we can clearly observe Choco returning null due to a timeout.

Additionally, by modifying the following properties in the TestChocoIntRangeAndTimeout.jpf file:

symbolic.min_int = -21474836
symbolic.max_int = 21474836

to 

symbolic.min_int=-2147483648
symbolic.max_int=2147483647

we can trigger an IllegalArgumentException. Note that it's not necessary to use the full Integer.MIN_VALUE or Integer.MAX_VALUE; any values significantly beyond ±21474836 are enough to reproduce the exception.

sohah commented on jpf-symbc/src/examples/TestChocoIntRangeAndTimeout.java:
```diff
@@ -0,0 +1,18 @@
+import org.sosy_lab.sv_benchmarks.Verifier;
+
+public class TestChocoIntRangeAndTimeout {
+    public static void main(String[] args) {
+        int x = Verifier.nondetInt();
+        test(x);
+    }
+
+    public static void test(int x) {
+        int z = x + x;
+        if(x < z) {
+            assert false;
+            System.out.println("Branch F001");
+        } else {
+            System.out.println("Branch F002");
+        }
+    }
+}
```
Can we have these tests and their corresponding configuration files under
examples/solvers/choco. You'll probablly add more solvers and more tests as we progress.

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/numeric/solvers/ProblemChoco.java:
```diff
@@ -324,8 +335,12 @@ public Boolean solve() {
-//        if (result == null)
- //       	System.out.println("Choco PC"+pb.pretty());
+
+		if (result == null) {
+			throw new RuntimeException("# Error: Choco returned null (possibly due to timeout).\n" +
+					"Time limit: " + ProblemChoco.timeBound + "\n" +
+					"Problem state:\n" + pb.pretty());
+		}
```
It is okay to raise an exception here, but as we discussed we should catch that when we have multiple solvers, we should not crash, but rather count on other solvers, i.e., z3 to give us an answer.

sohah commented:
Thanks Rehan for your contribution. I think we can accept this pull request after you make some minor changes.

sohah commented:
> @sohah I have added the source reference, Its a PDF, so I was not able to point to a specific page number.
> 
> I have also added an example: TestChocoIntRangeAndTimeout. In this test, we can clearly observe Choco returning null due to a timeout.
> 
> Additionally, by modifying the following properties in the TestChocoIntRangeAndTimeout.jpf file:
> 
> symbolic.min_int = -21474836 symbolic.max_int = 21474836
> 
> to
> 
> symbolic.min_int=-2147483648 symbolic.max_int=2147483647
> 
> we can trigger an IllegalArgumentException. Note that it's not necessary to use the full Integer.MIN_VALUE or Integer.MAX_VALUE; any values significantly beyond ±21474836 are enough to reproduce the exception.

Also, in the pdf, isn't that comment about the bounds is on page 6? Can we have this in the comment?

RehanChalana committed:
refactor: move Choco test files to solvers/choco directory

- Moved TestChocoIntRangeAndTimeout.java and .jpf to solvers/choco/

- Improved reference comment for Choco's integer overflow issue (added page number from user guide)

RehanChalana commented:
Thanks for the review!

I have moved TestChocoIntRangeAndTimeout.java and TestChocoIntRangeAndTimeout.jpf to examples/solvers/choco directory.

I have also improved the reference comment for the choco documentation including the page number.

sohah commented:
@RehanChalana , thank you for your contribution.
@yannicnoller , I think this is good to be accepted.
